controller/kubernetes_controller.go

```diff
@@ -0,0 +1,409 @@
+	if err == nil {
+		kc.queue.Forget(key)
+		return
+	}
+
+	if kc.queue.NumRequeues(key) < maxRetries {
+		logrus.Warnf("Error syncing Longhorn volume status %v: %v", key, err)
+		kc.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logrus.Warnf("Dropping Longhorn volume %v out of the queue: %v", key, err)
+	kc.queue.Forget(key)
+}
+
+func (kc *KubernetesController) syncKubernetesStatus(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "fail to sync %v", key)
+	}()
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	err = kc.cleanupForPVDeletion(name)
+	if err != nil {
+		return err
+	}
+
+	pv, err := kc.kubeClient.CoreV1().PersistentVolumes().Get(name, metav1.GetOptions{})
+	if err != nil && !IsErrorNotFound(err) {
+		return errors.Wrapf(err, "Error getting Persistent Volume %s", name)
+	}
+	volumeName, err := kc.validatePersistentVolume(pv)
+	if err != nil {
+		return err
+	}
+	if volumeName == "" {
+		return nil
+	}
+
+	volume, err := kc.ds.GetVolume(volumeName)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			logrus.Infof("Longhorn volume %v has been deleted", volumeName)
+		}
+		return nil
+	}
+
+	// existing volume may be reused by other pv
+	ks := &volume.Status.KubernetesStatus
+	if ks.PVName != name {
+		ks.Namespace = ""
+		ks.PVCName = ""
+		ks.PodName = ""
+		ks.PodStatus = ""
+		ks.WorkloadName = ""
+		ks.WorkloadType = ""
+		ks.LastPVCRefAt = ""
+		ks.LastPodRefAt = ""
+	}
+
+	// PVC gets bound with PV when PVStatus becomes Bound
+	if pv.Spec.ClaimRef != nil && pv.Status.Phase == v1.VolumeBound && ks.PVStatus != string(pv.Status.Phase) {
+		ks.PVCName = pv.Spec.ClaimRef.Name
+		ks.Namespace = pv.Spec.ClaimRef.Namespace
+		ks.LastPVCRefAt = ""
+	}
+
+	// PVC is no longer bound with PV. indicating history data by setting <LastPVCRefAt>
+	if ks.PVStatus == string(v1.VolumeBound) && ks.PVStatus != string(pv.Status.Phase) {
+		ks.LastPVCRefAt = util.Now()
+	}
+
+	var pod *v1.Pod
+	pod, err = kc.getMountPod(ks.Namespace, ks.PVCName)
+	if err != nil {
+		return err
+	}
+
+	// pod deleted. indicating history data by setting <LastPodRefAt>
+	if ks.PodName != "" && ks.LastPodRefAt == "" && (pod == nil || pod.DeletionTimestamp != nil) {
+		if ks.LastPodRefAt == "" {
+			ks.LastPodRefAt = util.Now()
+		}
+	}
+	if pod != nil {
+		workloadName, workloadType := kc.detectWorkload(pod, ks)
+		ks.PodName = pod.Name
+		ks.PodStatus = string(pod.Status.Phase)
+		ks.WorkloadName = workloadName
+		ks.WorkloadType = workloadType
+		ks.LastPodRefAt = ""
+	}
+
+	// if pv.Status.Phase == v1.VolumePending || pv.Status.Phase == v1.VolumeAvailable, do nothing
+	ks.PVName = name
+	ks.PVStatus = string(pv.Status.Phase)
+
+	volume, err = kc.ds.UpdateVolume(volume)
+	if err != nil {
+		return err
+	}
+	kc.eventRecorder.Eventf(volume, v1.EventTypeNormal, EventReasonStart, "Updating KubernetesStatus for volume %v", volume.Name)
```

Comment:
Now, this event will be out there a few times per second. See `kubectl -n longhorn-system get events`

Also, you won't want to send the user a message every time when status updated.